### DISCUSSION TITLE
Support test functions in same file

### DISCUSSION THREAD

**Discussion by ntrel** [vlang/v#11578] [Ideas]
https://vlang.io/docs#testing:
> All test functions have to be placed in *_test.v files
> To run the tests do v hello_test.v

To encourage better maintenance, V could support defining a test function following a symbol in the same file. Otherwise the coder will forget (or not bother) to update the test, because they don't see it when working on a file.

Then to run tests in a file, use something like `v -test hello.v`.

Edit: This is also important for testing examples in documentation, see below.

**Comment by medvednikov** [comment]
I understand and I agree.

The problem is that very often one function can have dozens of test functions. They should not pollute the code base.

Perhaps `vfmt` should cut and paste all `fn test_*` into newly created `*_test.v` files. 

This makes creating new tests easier, but doesn't solve the problem of not updated tests.

**Comment by PavelVozenilek** [comment]
It is advantageous to have both options. Some functionality is easy to test and the tests could serve as handy examples of use. Those are best placed after the relevant functions.

Really extensive testing takes _a lot_ of space and makes navigation in source file clumsy. It is better to move it out of sight.

In some rare situations it may be even useful to place short test _inside_ the regular code, to give a hint what part of the code is being tested. (I believe such functionality could be implemented in Nim language.)



----------------

I once thought about the following feature:


```
-- source code --
fn foo() {}  // some function

TEST() // small test for the foo function, unnamed
{
}

TEST() // another small foo test
{
}

TEST(complicated-foo-test1); // this is like forward, indication that the test is placed in accompanying test file. Has to be named.

TEST(complicated-foo-test2); // another test moved to test file

```

plus 

```
--- the accompanying test file --
// all forward tests have to be here, in the same order as they are listed in source file
TEST(complicated-foo-test1) 
{
}

TEST(complicated-foo-test2)
{
}

```


This makes all the tests known, tests which are not supposed to be read over and over are moved aside, navigation is easy, nothing can get lost.  The names would be unique only per file, not per project.

**Comment by ntrel** [comment]
> It is advantageous to have both options
> Some functionality is easy to test and the tests could serve as handy examples of use

Absolutely, in D you can add a `///` doc-comment to a `unittest` block following a symbol, and the test appears in the generated docs as example code. This is very useful to make sure the documentation examples compile and work correctly.

> Really extensive testing takes a lot of space and makes navigation in source file clumsy

Yes, good point. Running `v --test foo.v` could also run the tests in `foo_test.v`. (BTW my editor supports code folding, so I can collapse the test function bodies quite easily. It would be possible for editors to automatically collapse test blocks too, via a plugin or script). But putting the less interesting bulk tests in a separate file is best.

> it may be even useful to place short test inside the regular code

`assert` inside normal functions is supported, but maybe you meant something else.

**Comment by PavelVozenilek** [comment]
> > it may be even useful to place short test inside the regular code

> assert inside normal functions is supported, but maybe you meant something else.


I meant this:

```
fn foo()
{
   if some-rare-condition {
      ...
      TEST()
       {
          ... // short test here, its placement indicates what it is testing
       }
       ...
   }
   ...
}

```
Not something for common use, only as option for certain situations.

**Comment by watzon** [comment]
Gonna add my own two cents here. Personally I love the idea of having tests in documentation, as that forces people to write documentation to test their code, and doesn't force you to maintain separate `*_test.v` files. We could potentially have something like:

```v
// Returns foo
// @example
//     foo := foo()
//     assert foo == 'foo'
fn foo() string {
  return 'foo'
}
```

**Comment by ntrel** [comment]
I think the example should be written as a function (rather than embedded in a comment where there's no syntax highlighting), e.g.:
```v
[example]
fn foo_test(){
  f := foo()
  assert f == 'foo'
}
```

**Comment by watzon** [comment]
No reason examples in comments couldn't have syntax highlighting. Pretty easy to do in a syntax definition actually.
